From d665f4436cff7ce477bc144a6eb534125fff5349 Mon Sep 17 00:00:00 2001
From: Max Kasperowski <mka@informatik.uni-kiel.de>
Date: Tue, 17 Oct 2023 11:02:22 +0200
Subject: [PATCH 5/7] Add top-down property to nodeContext

Only consider existing node height in the top-down layout case
---
 .../common/nodespacing/internal/NodeContext.java   |  5 +++++
 .../internal/algorithm/NodeSizeCalculator.java     | 14 ++++++++++----
 .../elk/alg/layered/issues/Issue701Test.java       |  5 ++---
 3 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/NodeContext.java b/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/NodeContext.java
index a86770ef85..e69ad4b31c 100644
--- a/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/NodeContext.java
+++ b/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/NodeContext.java
@@ -86,6 +86,8 @@ public final class NodeContext {
     public final double portLabelSpacingVertical;
     /** Margin to leave around the set of ports on each side. */
     public final ElkMargin surroundingPortMargins;
+    /** Whether node is being laid out in top-down layout mode. */
+    public final boolean topdownLayout;
 
 
     /////////////////////////////////////////////////////////////////////////////////
@@ -131,6 +133,9 @@ public NodeContext(final GraphAdapter<?> parentGraph, final NodeAdapter<?> node)
         this.node = node;
         this.nodeSize = new KVector(node.getSize());
         
+        // Top-down layout
+        topdownLayout = node.getProperty(CoreOptions.TOPDOWN_LAYOUT);
+        
         // Compound node
         treatAsCompoundNode = node.isCompoundNode() || node.getProperty(CoreOptions.INSIDE_SELF_LOOPS_ACTIVATE);
         
diff --git a/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeSizeCalculator.java b/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeSizeCalculator.java
index 0e063f7387..b03f1341a7 100644
--- a/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeSizeCalculator.java
+++ b/plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeSizeCalculator.java
@@ -48,8 +48,11 @@ public static void setNodeWidth(final NodeContext nodeContext) {
         } else {
             // Ask the cell system how wide it would like to be or take the node's width if it has already been set to a
             // greater value
-            width = Math.max(nodeSize.x, nodeContext.nodeContainer.getMinimumWidth());
-//            width = nodeContext.nodeContainer.getMinimumWidth();
+            if (nodeContext.topdownLayout) {
+                width = Math.max(nodeSize.x, nodeContext.nodeContainer.getMinimumWidth());
+            } else {
+                width = nodeContext.nodeContainer.getMinimumWidth();
+            }
             
             // If we include node labels and outside node labels are not to overhang, we need to include those as well
             if (nodeContext.sizeConstraints.contains(SizeConstraint.NODE_LABELS)
@@ -99,8 +102,11 @@ public static void setNodeHeight(final NodeContext nodeContext) {
         } else {
             // Ask the cell system how high it would like to be or take the node's height if it has already been set to 
             // a greater value
-            height = Math.max(nodeSize.y, nodeContext.nodeContainer.getMinimumHeight());
-//            height = Math.max(0, nodeContext.nodeContainer.getMinimumHeight());
+            if (nodeContext.topdownLayout) {
+                height = Math.max(nodeSize.y, nodeContext.nodeContainer.getMinimumHeight());
+            } else {
+                height = nodeContext.nodeContainer.getMinimumHeight();
+            }
             
             // If we include node labels and outside node labels are not to overhang, we need to include those as well
             if (nodeContext.sizeConstraints.contains(SizeConstraint.NODE_LABELS)
diff --git a/test/org.eclipse.elk.alg.layered.test/src/org/eclipse/elk/alg/layered/issues/Issue701Test.java b/test/org.eclipse.elk.alg.layered.test/src/org/eclipse/elk/alg/layered/issues/Issue701Test.java
index d9abb6578e..a1ba0b3825 100644
--- a/test/org.eclipse.elk.alg.layered.test/src/org/eclipse/elk/alg/layered/issues/Issue701Test.java
+++ b/test/org.eclipse.elk.alg.layered.test/src/org/eclipse/elk/alg/layered/issues/Issue701Test.java
@@ -415,8 +415,7 @@ private void assertNodeWidthAccordingToPortWidth(ElkNode node, ElkPort port, Elk
             // org.eclipse.elk.alg.common.nodespacing.internal.algorithm.HorizontalPortPlacementSizeCalculator.calculateHorizontalNodeSizeRequiredByFreePorts(NodeContext,
             // PortSide)
             assertEquals("Wrong node width according to port(s) label width.",
-                    Math.max(spacingBetweenPort + Math.max(portLabelWidth, oppositePortLabelWidth) + spacingBetweenPort,
-                            node.getWidth()),
+                    spacingBetweenPort + Math.max(portLabelWidth, oppositePortLabelWidth) + spacingBetweenPort,
                     node.getWidth(), 0);
         } else {
             double spacingLabelPort = port.getProperty(CoreOptions.SPACING_LABEL_PORT_HORIZONTAL);
@@ -482,7 +481,7 @@ private void assertNodeHeightAccordingToPortHeight(ElkNode node, ElkPort port, E
                                     + Math.max(spacingBetweenPort, nodePadding.getBottom()));
 
             assertEquals("Wrong node height according to port label height.",
-                    Math.max(portHeightPlusPadding, Math.max(spacingBetweenPort, node.getHeight())), node.getHeight(), 0);
+                    Math.max(portHeightPlusPadding, spacingBetweenPort), node.getHeight(), 0);
             assertEquals("Wrong node label location.", ElkUtil.absolutePosition(node).y + nodeLabelPadding.getTop(),
                     ElkUtil.absolutePosition(nodeLabel).y, 0);
         } else {

